From 5e1690bee7e86250e7970a56c1a4fc2075b448e1 Mon Sep 17 00:00:00 2001
From: nailyk-fr <nailyk_git@nailyk.fr>
Date: Tue, 8 Aug 2017 21:23:35 +0200
Subject: [PATCH 2/2] twrpapp: automatically exclude on AB devices

The TWRP app does not currently support AB devices.

 * If AB_OTA_UPDATER := true, then exclude TWRP app

 * To manually exclude TWRP app for non-AB devices,
   use BoardConfig flag: 'TW_EXCLUDE_TWRPAPP := true'

Original commit: https://gerrit.omnirom.org/#/c/android_bootable_recovery/+/27694/

Change-Id: I0aa1fb2ebe03ceb19c766178544039a82497cdd4
---
 Android.mk                     |  4 ++++
 data.cpp                       |  4 ++++
 gui/theme/common/landscape.xml |  1 +
 gui/theme/common/portrait.xml  |  1 +
 gui/theme/common/watch.xml     |  1 +
 prebuilt/Android.mk            | 18 ++++++++++--------
 6 files changed, 21 insertions(+), 8 deletions(-)

diff --git a/Android.mk b/Android.mk
index 57c300332..a75188ca1 100755
--- a/Android.mk
+++ b/Android.mk
@@ -199,6 +199,7 @@ ifeq ($(AB_OTA_UPDATER),true)
     LOCAL_CFLAGS += -DAB_OTA_UPDATER=1
     LOCAL_SHARED_LIBRARIES += libhardware android.hardware.boot@1.0
     TWRP_REQUIRED_MODULES += libhardware
+    TW_EXCLUDE_TWRPAPP := true
 endif
 
 LOCAL_MODULE_PATH := $(TARGET_RECOVERY_ROOT_OUT)/sbin
@@ -425,6 +426,9 @@ ifneq ($(TW_INCLUDE_LIBRESETPROP),)
         LOCAL_CFLAGS += -DTW_INCLUDE_LIBRESETPROP
     endif
 endif
+ifeq ($(TW_EXCLUDE_TWRPAPP),true)
+    LOCAL_CFLAGS += -DTW_EXCLUDE_TWRPAPP
+endif
 
 TWRP_REQUIRED_MODULES += \
     dump_image \
diff --git a/data.cpp b/data.cpp
index e6f1ac0ce..bc3aff920 100755
--- a/data.cpp
+++ b/data.cpp
@@ -904,10 +904,14 @@ void DataManager::SetDefaultValues()
 	mConst.SetValue("tw_app_installed_in_system", "0");
 #else
 	mConst.SetValue("tw_oem_build", "0");
+  #ifdef TW_EXCLUDE_TWRPAPP
+	mConst.SetValue("tw_app_prompt", "-1");
+  #else
 	mPersist.SetValue("tw_app_prompt", "1");
 	mPersist.SetValue("tw_app_install_system", "1");
 	mData.SetValue("tw_app_install_status", "0"); // 0 = no status, 1 = not installed, 2 = already installed
 	mData.SetValue("tw_app_installed_in_system", "0");
+  #endif
 #endif
 
 	mData.SetValue("tw_enable_adb_backup", "0");
diff --git a/gui/theme/common/landscape.xml b/gui/theme/common/landscape.xml
index daee7f337..36a8b701f 100755
--- a/gui/theme/common/landscape.xml
+++ b/gui/theme/common/landscape.xml
@@ -3015,6 +3015,7 @@
 					<data variable="tw_military_time"/>
 				</listitem>
 				<listitem name="{@reboot_install_app_prompt_install=Prompt to install TWRP app on every reboot}">
+					<condition var1="tw_app_prompt" op="!=" var2="-1"/>
 					<data variable="tw_app_prompt"/>
 				</listitem>
 				<listitem name="{@simact_chk=Simulate actions for theme testing}">
diff --git a/gui/theme/common/portrait.xml b/gui/theme/common/portrait.xml
index bead7e619..86eb7806a 100755
--- a/gui/theme/common/portrait.xml
+++ b/gui/theme/common/portrait.xml
@@ -3227,6 +3227,7 @@
 					<data variable="tw_samsung_navbar"/>
 				</listitem>
 				<listitem name="{@reboot_install_app_prompt_install=Prompt to install TWRP app on every reboot}">
+					<condition var1="tw_app_prompt" op="!=" var2="-1"/>
 					<data variable="tw_app_prompt"/>
 				</listitem>
 				<listitem name="{@simact_chk=Simulate actions for theme testing}">
diff --git a/gui/theme/common/watch.xml b/gui/theme/common/watch.xml
index 74212030d..ee58e70d7 100755
--- a/gui/theme/common/watch.xml
+++ b/gui/theme/common/watch.xml
@@ -3717,6 +3717,7 @@
 					<data variable="tw_military_time"/>
 				</listitem>
 				<listitem name="{@reboot_install_app_prompt_install=Prompt to install TWRP app on every reboot}">
+					<condition var1="tw_app_prompt" op="!=" var2="-1"/>
 					<data variable="tw_app_prompt"/>
 				</listitem>
 				<listitem name="{@simact_chk=Simulate actions for theme testing}">
diff --git a/prebuilt/Android.mk b/prebuilt/Android.mk
index a4f6f2e4b..9f6f1653d 100644
--- a/prebuilt/Android.mk
+++ b/prebuilt/Android.mk
@@ -562,14 +562,16 @@ ifeq ($(TW_USE_TOOLBOX), true)
     include $(BUILD_PREBUILT)
 endif
 
-#TWRP App "placeholder"
-include $(CLEAR_VARS)
-LOCAL_MODULE := me.twrp.twrpapp.apk
-LOCAL_MODULE_TAGS := eng
-LOCAL_MODULE_CLASS := RECOVERY_EXECUTABLES
-LOCAL_MODULE_PATH := $(TARGET_RECOVERY_ROOT_OUT)/sbin
-LOCAL_SRC_FILES := $(LOCAL_MODULE)
-include $(BUILD_PREBUILT)
+ifneq ($(TW_EXCLUDE_TWRPAPP),true)
+	#TWRP App "placeholder"
+	include $(CLEAR_VARS)
+	LOCAL_MODULE := me.twrp.twrpapp.apk
+	LOCAL_MODULE_TAGS := eng
+	LOCAL_MODULE_CLASS := RECOVERY_EXECUTABLES
+	LOCAL_MODULE_PATH := $(TARGET_RECOVERY_ROOT_OUT)/sbin
+	LOCAL_SRC_FILES := $(LOCAL_MODULE)
+	include $(BUILD_PREBUILT)
+endif
 
 #TWRP App permissions for Android 9+
 include $(CLEAR_VARS)
-- 
2.27.0

